bridges/snowbridge/pallets/system-v2/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+}
+
+#[cfg(feature = "runtime-benchmarks")]
+pub trait BenchmarkHelper<O>
+where
+	O: OriginTrait,
+{
+	fn make_xcm_origin(location: Location) -> O;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::dispatch::RawOrigin;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Send messages to Ethereum
+		type OutboundQueue: SendMessage;
+
+		/// Origin check for XCM locations that can create agents
+		type SiblingOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
+
+		/// Converts Location to AgentId
+		type AgentIdOf: ConvertLocation<AgentId>;
+
+		type WeightInfo: WeightInfo;
+
+		/// This chain's Universal Location.
+		type UniversalLocation: Get<InteriorLocation>;
+
+		// The bridges configured Ethereum location
+		type EthereumLocation: Get<Location>;
+
+		#[cfg(feature = "runtime-benchmarks")]
+		type Helper: BenchmarkHelper<Self::RuntimeOrigin>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An CreateAgent message was sent to the Gateway
+		CreateAgent { location: Box<Location>, agent_id: AgentId },
+		/// Register Polkadot-native token as a wrapped ERC20 token on Ethereum
+		RegisterToken {
+			/// Location of Polkadot-native token
+			location: VersionedLocation,
+			/// ID of Polkadot-native token on Ethereum
+			foreign_token_id: H256,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		LocationConversionFailed,
+		AgentAlreadyCreated,
+		NoAgent,
+		UnsupportedLocationVersion,
+		InvalidLocation,
+		Send(SendError),
+	}
+
+	/// The set of registered agents
+	#[pallet::storage]
+	#[pallet::getter(fn agents)]
+	pub type Agents<T: Config> = StorageMap<_, Twox64Concat, AgentId, (), OptionQuery>;
+
+	/// Lookup table for foreign token ID to native location relative to ethereum
+	#[pallet::storage]
+	pub type ForeignToNativeId<T: Config> =
+		StorageMap<_, Blake2_128Concat, TokenId, xcm::v5::Location, OptionQuery>;
+
+	/// Lookup table for native location relative to ethereum to foreign token ID
+	#[pallet::storage]
+	pub type NativeToForeignId<T: Config> =
+		StorageMap<_, Blake2_128Concat, xcm::v5::Location, TokenId, OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Sends a command to the Gateway contract to instantiate a new agent contract representing
+		/// `origin`.
+		///
+		/// Fee required: Yes
+		///
+		/// - `origin`: Must be `Location` of a sibling parachain
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::create_agent())]
+		pub fn create_agent(
+			origin: OriginFor<T>,
+			location: Box<VersionedLocation>,
+			fee: u128,
+		) -> DispatchResult {
+			let _ = match origin.clone().into() {
+				Ok(RawOrigin::Root) => Ok(Here.into()),
```

Comment:
@yrong ditto, we I think we should force root to go through AH.

Reply:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/33a8356e12615d75e3bcd0a1b302c468adeb612f